src/storage-controller/src/storage_usage.rs

```diff
@@ -0,0 +1,208 @@
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Utilities for working with the special `mz_storage_usage_by_shard` builtin
+//! table.
+
+use std::fmt::Display;
+use std::sync::Arc;
+use std::time::Duration;
+
+use differential_dataflow::lattice::Lattice;
+use mz_ore::now::EpochMillis;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::cfg::USE_CRITICAL_SINCE_SNAPSHOT;
+use mz_persist_client::read::ReadHandle;
+use mz_persist_client::Diagnostics;
+use mz_persist_types::codec_impls::UnitSchema;
+use mz_persist_types::Codec64;
+use mz_repr::{Diff, GlobalId, Row, TimestampManipulation};
+use mz_storage_client::client::TimestamplessUpdate;
+use mz_storage_types::controller::{CollectionMetadata, StorageError};
+use mz_storage_types::sources::SourceData;
+use mz_timestamp_oracle::TimestampOracle;
+use mz_txn_wal::txn_read::TxnsRead;
+use timely::order::TotalOrder;
+use timely::progress::Timestamp;
+use tracing::debug;
+
+use crate::persist_handles;
+
+#[allow(unused)]
+pub struct PruneStorageUsageTask<
+    T: Timestamp + Lattice + Codec64 + From<EpochMillis> + TimestampManipulation,
+> {
+    /// [GlobalId] of our `mz_storage_usage_by_shard` collection. Mostly for
+    /// diagnostics/logging.
+    pub collection_id: GlobalId,
+
+    /// [CollectionMetadata] of our `mz_storage_usage_by_shard` collection.
+    pub collection_metadata: CollectionMetadata,
+
+    /// Configured retention period.
+    pub retention_period: Duration,
+
+    /// Handle to shared persist client cache.
+    pub persist: Arc<PersistClientCache>,
+
+    /// Process-global write lock.
+    ///
+    /// NOTE(aljoscha): I wish we didn't have this, but here we are. We can get
+    /// rid of this when we replace the write lock by, say, optimistic
+    /// concurrency control.
+    pub write_lock: Arc<tokio::sync::Mutex<()>>,
+
+    /// Write handle for table shards.
+    pub persist_table_worker: persist_handles::PersistTableWriteWorker<T>,
+
+    /// Handle to a shared [TxnsRead].
+    pub txns_read: TxnsRead<T>,
+
+    /// Timestamp oracle for the table timeline.
+    pub timestamp_oracle: Arc<dyn TimestampOracle<T> + Send + Sync>,
+}
+
+impl<T> PruneStorageUsageTask<T>
+where
+    T: Timestamp
+        + Lattice
+        + TotalOrder
+        + Codec64
+        + From<EpochMillis>
+        + TimestampManipulation
+        + Display,
+    u128: From<T>,
+{
+    #[allow(unused)]
+    pub async fn run_once(mut self) {
+        // We retry only until we succeed pruning once.
+        loop {
+            let read_ts = self.timestamp_oracle.read_ts().await;
+
+            // TODO: Instead of snapshotting fresh every time, we could
+            // continually listen while we attempt to get our retractions in at
+            // the next write timestamp.
```

Comment:
I think this is probably a requirement, otherwise we'll get starved out by whoever is advancing the upper.